src/checkers/inference/BaseInferenceResult.java

```diff
@@ -0,0 +1,39 @@
+package checkers.inference;
+
+import checkers.inference.model.Constraint;
+
+import javax.lang.model.element.AnnotationMirror;
+import java.util.Collection;
+import java.util.Map;
+
+public class BaseInferenceResult implements InferenceResult {
+
+    protected final Map<Integer, AnnotationMirror> inferredResults;
+    protected final Collection<Constraint> unsatisfiableConstraints;
+
+    public BaseInferenceResult(Map<Integer, AnnotationMirror> inferredResults,
+                               Collection<Constraint> unsatisfiableConstraints) {
+        this.inferredResults = inferredResults;
+        this.unsatisfiableConstraints = unsatisfiableConstraints;
+    }
+
+    @Override
+    public boolean doesVariableExist(int varId) {
+        return inferredResults.containsKey(varId);
+    }
+
+    @Override
+    public AnnotationMirror getAnnotation(int varId) {
+        return inferredResults.get(varId);
+    }
+
+    @Override
+    public boolean hasSolution() {
+        return inferredResults != null;
```

Comment:
What is the purpose of this function `hasSolution()`?

Could you clarify the corner case: an empty `inferredResults` is "an empty solution" or represents "no solution"?

Reply:
Also, using `null` state represents "no solution" might not be the best option, as in this case the field `inferredResult` is nullable, which means using methods `getAnnotation(varId)` and `doesVariableExist(varId)` has the risk of potential NPE.

Reply:
Purpose of `hasSolution()` is to let us know if the inference has solution or not. So that if there isn't, explanation gives us the reason why constraints are not solvable.
`inferredResults` is empty doesn't mean there is no solution. I tried using empty `inferredResults` to represent no solution, but got error: some trivial testcases in CFI doesn't have any `VarId` in its solution thus has an empty map of `inferredResults`. If explanation feature sees there is empty map, and starts explanation, it got an `AssertionError` because explanation engine asserts constraints are not solvable but actually it is.
So, we can't use empty map to represent no solution case. 
We could tell in the Javadoc that before calling `getAnnotation(varId)` and `doesVariableExist(varId)`, users need to check if there is solution or not using `hasSolution()`. Throwing NPE makes sense if you try to get a solution for an id if there is no solution.

Reply:
Sure, make sense. Thanks for the explanation!